graphql/schemabuilder/function.go

```diff
@@ -111,6 +115,48 @@ func (sb *schemaBuilder) buildFederatedFunction(typ reflect.Type, m *method) (*g
 	return field, nil
 }
 
+func (sb *schemaBuilder) buildShadowObjectFederationFunction(typ reflect.Type, m *method) (*graphql.Field, error) {
+	funcCtx := &funcContext{typ: typ}
+
+	// Create the input type in the format of struct{keys: *federationtype}
+	shadowType := reflect.TypeOf(m.FederationType)
+	input := reflect.StructOf([]reflect.StructField{
+		{
+			Name: "Keys",
+			Type: reflect.SliceOf(shadowType),
+		},
+	})
+	i := make([]reflect.Type, 1)
+	i[0] = input
+
+	argParser, argType, _, err := funcCtx.getArgParserAndTyp(sb, i)
+	if err != nil {
+		return nil, err
+	}
+	funcCtx.hasArgs = argParser != nil
+	args, err := funcCtx.argsTypeMap(argType)
+	if err != nil {
+		return nil, err
+	}
+	returnType, err := sb.getType(reflect.TypeOf(m.FederationType))
```

Comment:
I think this should be something like `rType := &graphql.NonNull{Type: &graphql.List{Type: retType}}`
unless thunder will think the return type is `device{}` rather than `[]*device` and then pass the entire slice rather than individual element to the downstream field func.